### ISSUE TITLE
BUG: Trace returns results for non-square matrices

### ISSUE THREAD

**Issue by rot8** [numpy/numpy#25391]
### Describe the issue:

From a mathematical standpoint the trace is [only defined for square matrices ](https://en.wikipedia.org/wiki/Trace_(linear_algebra)) or more generally for matrices in ℝ<sup>n</sup>.

However, the [numpy implementation](https://github.com/numpy/numpy/blob/v1.26.0/numpy/core/fromnumeric.py#L1700-L1761) returns the sum of entries where all indices are the same which for square matrices coincides with the trace but also returns wrong results disguised as trace if the matrix is not square. This can result in unwanted behavior as one could use the trace to check if the matrix is square, expecting an error if it is not.

I can see that this implementation of the trace may have its use cases. But I would strongly suggest modifying the default function to return only the mathematically correct trace and extending the function with a flag which allows non-square matrices.

```
import numpy as np

A = np.array([[1, 2, 3],
              [1, 2, 3]])

print(A.trace())
```
Result: `3`
Expected Result: `"AssertionError: Trace is not defined for non-square matrices"`

Further, I am interested if anyone knows why this implementation was chosen and what is actual use cases are.

### Reproduce the code example:

```python
import numpy as np

A = np.array([[1, 2, 3],
              [1, 2, 3]])

print(A.trace())
```

### Error message:

```shell
N/A
```

### Python and NumPy Versions:

numpy: 1.21.5
sys: 3.9.13

**Comment by bgmwaring** [comment]
I agree, the trace of a non-square matrix should not be defined. I don't think it is helpful for a scientific tool to allow this in any way, it gives legitimacy to an incorrect interpretation of the function. If there is a use for the way trace() handles non-square arrays, then I cannot think of it.

**Comment by linus-md** [comment]
I could work on this but I could imagine, that this hurts backwards compatibility.

**Comment by rot8** [comment]
I can definitely help you if you want. But do you know how one can get an idea if this change would be accepted by a majority?

**Comment by linus-md** [comment]
We should see what the maintainers say.

**Comment by rkern** [comment]
Changes like this should be [proposed on the mailing list](https://mail.python.org/mailman/listinfo/numpy-discussion) to ensure that it gets attention.

But I doubt that we will introduce a new exception for mathematical purity.

> This can result in unwanted behavior as one could use the trace to check if the matrix is square, expecting an error if it is not.

Since the straightforward way to do this check is to examine the `.shape`, I don't think we will consider this as a use case for `trace()`.

**Comment by rot8** [comment]
Thank you very much, I will propose it.

Don't get me wrong, I think the function can be very useful to compute specific diagonal sums. But the function does not do what the trace actually should do. 

Further, I noticed that `np.array([1]).trace()` returns ` diag requires an array of at least two dimensions` while the trace of a scalar should return the scalar itself.
